src/Datadog.Trace.ClrProfiler.Managed/Integrations/MongoDbIntegration.cs

```diff
@@ -0,0 +1,181 @@
+using System;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Datadog.Trace.ClrProfiler.Integrations
+{
+    /// <summary>
+    /// Tracing integration for MongoDB.Driver.Core.
+    /// </summary>
+    public static class MongoDbIntegration
+    {
+        internal const string OperationName = "mongodb.query";
+        internal const string ServiceName = "mongodb";
+
+        /// <summary>
+        /// Wrap the original method by adding instrumentation code around it.
+        /// </summary>
+        /// <param name="wireProtocol">The IWireProtocol`1 instance we are replacing.</param>
+        /// <param name="connection">The connection.</param>
+        /// <param name="cancellationTokenSource">A cancellation token source.</param>
+        /// <returns>The original method's return value.</returns>
+        [InterceptMethod(
+            TargetAssembly = "MongoDB.Driver.Core",
+            TargetType = "MongoDB.Driver.Core.WireProtocol.IWireProtocol")]
+        [InterceptMethod(
+            TargetAssembly = "MongoDB.Driver.Core",
+            TargetType = "MongoDB.Driver.Core.WireProtocol.IWireProtocol`1")]
+        public static object Execute(object wireProtocol, object connection, object cancellationTokenSource)
```

Comment:
Why is our `Execute` function taking a `CancellationTokenSource` as argument and the original `Execute` with a `CancellationToken`? Shouldn't the signatures match?

Reply:
Good catch! Funny story about those... The CLR does some special magic with `CancellationTokenSource` and `CancellationToken`. If a method argument is of type `CancellationToken` (struct), but the method signature expects an `object`, the CLR _automatically_ wraps the token in a token source. No boxing required! Furthermore, if the token is `CancellationToken.None`, the passes `null` instead.

We found that is was much easier to rely on this behavior because `object` is a well-known type in method signatures and doesn't rely on metadata token. We already do this in all integrations that instrument async methods and have `CancellationToken` arguments.

Reply:
Interesting, thanks for the explanation 👍 Is this something that is described in the ECMA standard or just a behavior of Microsoft's CLR implementation?